web-client/src/main/java/no/nav/familie/webflux/filter/BearerTokenFilter.kt

```diff
@@ -0,0 +1,205 @@
+    request: ClientRequest,
+    function: ExchangeFunction,
+    grantType: OAuth2GrantType? = null
+): Mono<ClientResponse> {
+    val accessToken = genererAccessToken(
+        request,
+        clientConfigurationProperties,
+        oAuth2AccessTokenService,
+        grantType
+    )
+    return addBearerToken(request, function, accessToken)
+}
+
+private fun addBearerToken(
+    request: ClientRequest,
+    function: ExchangeFunction,
+    accessToken: String
+): Mono<ClientResponse> {
+    val modifiedRequest = ClientRequest.from(request).header(
+        "Authorization",
+        "Bearer " + accessToken
+    ).build()
+    return function.exchange(modifiedRequest)
+}
+
+private fun genererAccessToken(
+    request: ClientRequest,
+    clientConfigurationProperties: ClientConfigurationProperties,
+    oAuth2AccessTokenService: OAuth2AccessTokenService,
+    grantType: OAuth2GrantType? = null
+): String {
+    val clientProperties = clientPropertiesFor(
+        request.url(),
+        clientConfigurationProperties,
+        grantType
+    )
+    return oAuth2AccessTokenService.getAccessToken(clientProperties).accessToken
+}
+
+/**
+ * Finds client property for grantType if specified.
+ *
+ * If the grantType isn't specified:
+ *  - Returns first client property, if there is only one
+ *  - Returns client property for client_credentials or jwt_bearer
+ */
+private fun clientPropertiesFor(
+    uri: URI,
+    clientConfigurationProperties: ClientConfigurationProperties,
+    grantType: OAuth2GrantType?
+): ClientProperties {
+    val clientProperties = filterClientProperties(clientConfigurationProperties, uri)
+    return if (grantType == null) {
+        if (clientProperties.size == 1) {
+            clientProperties.first()
+        } else {
+            clientPropertiesForGrantType(clientProperties, clientCredentialOrJwtBearer(), uri)
+        }
+    } else {
+        clientPropertiesForGrantType(clientProperties, grantType, uri)
+    }
+}
+
+private fun filterClientProperties(
+    clientConfigurationProperties: ClientConfigurationProperties,
+    uri: URI
+) = clientConfigurationProperties
+    .registration
+    .values
+    .filter { uri.toString().startsWith(it.resourceUrl.toString()) }
+
+private fun clientPropertiesForGrantType(
+    values: List<ClientProperties>,
+    grantType: OAuth2GrantType,
+    uri: URI
+): ClientProperties {
+    return values.firstOrNull { grantType == it.grantType }
+        ?: error("could not find oauth2 client config for uri=$uri and grant type=$grantType")
+}
+
+private fun clientCredentialOrJwtBearer() =
+    if (erSystembruker()) OAuth2GrantType.CLIENT_CREDENTIALS else OAuth2GrantType.JWT_BEARER
+
+private fun erSystembruker(): Boolean {
+    return try {
+        val preferred_username =
+            SpringTokenValidationContextHolder().tokenValidationContext.getClaims("azuread")["preferred_username"]
+        return preferred_username == null
+    } catch (e: Throwable) {
```

Comment:
Kan vi bruke noko meir spesifikt her? Liker dårleg å måtte catche Throwable

Reply:
Yep 👍 Vet ikke helt hvorfor det ble endret til Throwable ut fra denne: https://github.com/navikt/familie-felles/pull/67
Jeg prøvd å holde meg borte fra runCatching pga at den catcher Throwable, men her er en intressant tråd om det:
https://stackoverflow.com/questions/63663998/why-does-kotlin-result-catch-throwables/64323675#64323675
Som lenker til https://twitter.com/relizarov/status/1299248297849303040

Reply:
Oi, denne var interessant! Det forklarer jo korfor det kjens som Kotlin legg opp til meir bruk av Throwable enn eg er vant til frå Java.

Kjøper ikkje heilt resonnementet hans heilt utan vidare, synes det er lovleg kort veg frå (tilsynelatande) anekdotisk argumentasjon til kontrær konklusjon, men det kan jo sjølvsagt vera mykje meir bak enn det som kjem fram i eit par korte twitter-meldingar. Så hadde vore interessant å sjå han utdjupe det meir, og det hadde også vore interessant med respons frå feks Brian Goetz i den tråden.

(Langt på sida av scopet for denne PR-en, sjølvsagt)